python/lsst/ts/criopy/salcomm/efd_cache.py

```diff
@@ -0,0 +1,552 @@
+            Ending interval, or None if timepoint is already cached.
+        """
+        half_duration = min_duration / 2
+
+        new_start = timepoint - TimeDelta(0.05, format="sec")
+        new_end = timepoint + min_duration
+
+        if self.start is None or self.end is None:
+            return new_start, new_end
+        elif timepoint < self.start:
+            duration = self.start - timepoint
+            if duration >= max_span:
+                return new_start, new_end
+            if duration < min_duration:
+                return (
+                    self.start - min_duration,
+                    self.start,
+                )
+            return timepoint - half_duration, self.start
+        elif timepoint > self.end:
+            duration = timepoint - self.end
+            if duration >= max_span:
+                return new_start, new_end
+            if duration < min_duration:
+                return self.end, self.end + min_duration
+            return self.end, timepoint + half_duration
+        return None, None
+
+    @property
+    def empty(self) -> bool:
+        """
+        Returns true when the cache is empty.
+
+        Returns
+        -------
+        empty : bool
+            True if cache is empty.
+        """
+        return self.data is None or self.data.empty
+
+    def get(self) -> EfdTopic | None:
+        """
+        Retrieves current data. Mimics SAL Topic get method, but returns
+        EfdTopic.
+
+        Retuns
+        ------
+        data : EfdTopic
+            Current (as set with call to set_current_time) data.
+
+        See also
+        --------
+        lsst.ts.salobj.topics.ReadTopic.get()
+        """
+        return self.current_data
+
+    def set_current_time(self, timepoint: Time) -> None:
+        """
+        Sets current timepoint in historical data.
+
+        Parameters
+        ----------
+        timepoint : `Time`
+            The cache topic entry will be set to mimics SAL topic at the given
+            time.
+        """
+        if self.empty:
+            self.current_data = None
+            return
+
+        assert self.data is not None
+        timestr = timepoint.iso + "+00:00"
+
+        row = self.data.loc[:timestr].tail(n=1)
+        if row.empty:
+            self.current_data = None
+            return
+
+        changed = True
+
+        if self.current_data is not None:
+            if self.current_data.private_sndStamp == row["private_sndStamp"].item():
+                changed = False
+
+        self.current_data = EfdTopic(row, changed)
+
+    def clear(self) -> None:
+        """
+        Clear cached data.
+        """
+        self.data = None
+
+    def merge(self, data: pd.DataFrame) -> None:
+        """
+        Merge current date with a new EFD data frame.
+
+        Parameters
+        ----------
+        data : `pd.DataFrame`
+            Newly added data frame. Data
+        """
+        if self.data is None or self.data.empty:
```

Comment:
`if self.empty` because you defined a new `empty` property above.

But actually, this part of logic is a little strange if we consider the following:

```python
elif data.empty:
    return
```

If we do have the `data.empty`, the above `self.data = data` might get this empty data.

Therefore, I would suggest to do (I assume the value of `data.empty` is a `bool`):

```python
if self.empty and (not data.empty):
```